code/iaas/allocator/src/main/java/io/cattle/platform/allocator/service/AbstractAllocator.java

```diff
@@ -326,9 +342,28 @@ protected void close(Iterator<AllocationCandidate> iter) {
     protected void populateConstraints(AllocationAttempt attempt, AllocationLog log) {
         List<Constraint> constraints = attempt.getConstraints();
 
+        Instance instance = attempt.getInstance();
+        Long requestedHostId = null;
+        if (instance != null) {
+            requestedHostId = DataAccessor.fields(instance).withKey(InstanceConstants.FIELD_REQUESTED_HOST_ID).as(Long.class);
+        }
+
         for (AllocationConstraintsProvider provider : allocationConstraintProviders) {
-            provider.appendConstraints(attempt, log, constraints);
+            if (requestedHostId == null || provider.isCritical()) {
+                provider.appendConstraints(attempt, log, constraints);
+            }
         }
+        Collections.sort(constraints, new Comparator<Constraint>() {
+            @Override
+            public int compare(Constraint o1, Constraint o2) {
+                if (o1 == o2) return 0;
```

Comment:
Can you just explain what your doing here?  Are you sorting to get soft constraints last??